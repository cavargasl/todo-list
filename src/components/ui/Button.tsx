import { cn } from "@/utils"
import { cva, type VariantProps } from "class-variance-authority"

const buttonVariants = cva(
  "inline-flex items-center justify-center gap-2 text-nowrap rounded-lg font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50",
  {
    variants: {
      variant: {
        default:
          "border-2 border-foreground bg-background hover:border-rose-500",
        ghost: "text-gray-500 hover:bg-foreground hover:text-white",
      },
      size: {
        default: "h-[60px] p-3",
        small: "h-[40px] p-2",
        icon: "size-[40px] rounded px-1 py-2",
      },
    },
    defaultVariants: {
      variant: "default",
      size: "default",
    },
  }
)

type ButtonProps = VariantProps<typeof buttonVariants> &
  React.ButtonHTMLAttributes<HTMLButtonElement>

export default function Button({
  className,
  variant,
  size,
  ...props
}: ButtonProps) {
  return (
    <button
      className={cn(buttonVariants({ variant, size, className }))}
      {...props}
    />
  )
}
