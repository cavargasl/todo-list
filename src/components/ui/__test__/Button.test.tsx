import { render } from "@testing-library/react"

import Button from "../Button"

describe("Button", () => {
  it("renders correctly with default props", () => {
    const { container } = render(<Button />)
    expect(container).toMatchSnapshot()
  })

  it("renders correctly with custom className", () => {
    const { container } = render(
      <Button className="font-extrabold text-pink-600" />
    )
    expect(container).toMatchSnapshot()
  })

  it("renders correctly with variant 'ghost'", () => {
    const { container } = render(<Button variant="ghost" />)
    expect(container).toMatchSnapshot()
  })

  it("renders correctly with size 'small'", () => {
    const { container } = render(<Button size="small" />)
    expect(container).toMatchSnapshot()
  })

  it("renders correctly with size 'icon'", () => {
    const { container } = render(<Button size="icon" />)
    expect(container).toMatchSnapshot()
  })

  it("renders correctly with children", () => {
    const { container } = render(<Button>Button</Button>)
    expect(container).toMatchSnapshot()
  })

  it("renders correctly with disabled prop", () => {
    const { container } = render(<Button disabled />)
    expect(container).toMatchSnapshot()
  })

  it("renders correctly with variant 'ghost' and size 'small'", () => {
    const { container } = render(<Button variant="ghost" size="small" />)
    expect(container).toMatchSnapshot()
  })

  it("renders correctly with variant 'ghost' and size 'icon'", () => {
    const { container } = render(<Button variant="ghost" size="icon" />)
    expect(container).toMatchSnapshot()
  })

  it("renders correctly with variant 'ghost', size 'default', and disabled", () => {
    const { container } = render(<Button variant="ghost" disabled />)
    expect(container).toMatchSnapshot()
  })
})
