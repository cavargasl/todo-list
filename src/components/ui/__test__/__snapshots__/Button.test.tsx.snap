// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders correctly with children 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 text-nowrap rounded-lg font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 border-2 border-foreground bg-background hover:border-rose-500 h-[60px] p-3"
  >
    Button
  </button>
</div>
`;

exports[`Button > renders correctly with custom className 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 text-nowrap rounded-lg transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 border-2 border-foreground bg-background hover:border-rose-500 h-[60px] p-3 font-extrabold text-pink-600"
  />
</div>
`;

exports[`Button > renders correctly with default props 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 text-nowrap rounded-lg font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 border-2 border-foreground bg-background hover:border-rose-500 h-[60px] p-3"
  />
</div>
`;

exports[`Button > renders correctly with disabled prop 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 text-nowrap rounded-lg font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 border-2 border-foreground bg-background hover:border-rose-500 h-[60px] p-3"
    disabled=""
  />
</div>
`;

exports[`Button > renders correctly with size 'icon' 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 text-nowrap font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 border-2 border-foreground bg-background hover:border-rose-500 size-[40px] rounded px-1 py-2"
  />
</div>
`;

exports[`Button > renders correctly with size 'small' 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 text-nowrap rounded-lg font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 border-2 border-foreground bg-background hover:border-rose-500 h-[40px] p-2"
  />
</div>
`;

exports[`Button > renders correctly with variant 'ghost' 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 text-nowrap rounded-lg font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 text-gray-500 hover:bg-foreground hover:text-white h-[60px] p-3"
  />
</div>
`;

exports[`Button > renders correctly with variant 'ghost' and size 'icon' 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 text-nowrap font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 text-gray-500 hover:bg-foreground hover:text-white size-[40px] rounded px-1 py-2"
  />
</div>
`;

exports[`Button > renders correctly with variant 'ghost' and size 'small' 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 text-nowrap rounded-lg font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 text-gray-500 hover:bg-foreground hover:text-white h-[40px] p-2"
  />
</div>
`;

exports[`Button > renders correctly with variant 'ghost', size 'default', and disabled 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 text-nowrap rounded-lg font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 text-gray-500 hover:bg-foreground hover:text-white h-[60px] p-3"
    disabled=""
  />
</div>
`;
